tests/test_similar.py

```diff
@@ -0,0 +1,141 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+
+
+import contextlib
+import os
+import re
+import sys
+import warnings
+from io import StringIO
+from os.path import abspath, dirname, join
+from typing import Iterator, List, TextIO
+
+import pytest
+
+from pylint.lint import Run
+
+HERE = abspath(dirname(__file__))
+DATA = join(HERE, "regrtest_data", "duplicate_code")
+CLEAN_PATH = re.escape(dirname(dirname(__file__)) + os.path.sep)
+
+
+@contextlib.contextmanager
+def _patch_streams(out: TextIO) -> Iterator:
+    sys.stderr = sys.stdout = out
+    try:
+        yield
+    finally:
+        sys.stderr = sys.__stderr__
+        sys.stdout = sys.__stdout__
+
+
+class TestSimilarCodeChecker:
+    def _runtest(self, args: List[str], code: int) -> None:
+        """Runs the tests and sees if output code is as expected."""
+        out = StringIO()
+        pylint_code = self._run_pylint(args, out=out)
+        output = out.getvalue()
+        msg = f"expected output status {code}, got {pylint_code}"
+        if output is not None:
+            msg = f"{msg}. Below pylint output: \n{output}"
+        assert pylint_code == code, msg
+
+    @staticmethod
+    def _run_pylint(args: List[str], out: TextIO) -> int:
+        """Runs pylint with a patched output."""
+        args = args + ["--persistent=no"]
+        with _patch_streams(out):
+            with pytest.raises(SystemExit) as cm:
+                with warnings.catch_warnings():
+                    warnings.simplefilter("ignore")
+                    Run(args)
+            return cm.value.code
+
+    @staticmethod
+    def _clean_paths(output: str) -> str:
+        """Normalize path to the tests directory."""
+        output = re.sub(CLEAN_PATH, "", output, flags=re.MULTILINE)
+        return output.replace("\\", "/")
```

Comment:
I feel maybe using ``pathlib.Path`` would permit to remove this but I'm not certain of what the ``re.sub`` is doing.

Reply:
It removes the `Users/.../.../.../pylint/` element from the `output` so that we get a direct path to the test file. This is because in normal `pylint` output you get the full path name and we don't want that there because the path names changes depending on test environment (`Users/Daniel/...` vs. `Users/Pierre/...`)

Reply:
He, then we can use [os.path.relpath()](https://docs.python.org/3/library/os.path.html#os.path.relpath) or [PurePath.relative_to()](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.relative_to).

Reply:
Not really, since these aren't single strings.

```console
❯ pylint test_dir/another_dir/test_data/test.py
************* Module test
test_dir/another_dir/test_data/test.py:1:0: W0104: Statement seems to have no effect (pointless-statement)
test_dir/another_dir/test_data/test.py:2:0: W0104: Statement seems to have no effect (pointless-statement)
test_dir/another_dir/test_data/test.py:3:0: W0104: Statement seems to have no effect (pointless-statement)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

It is the output similar to this. However, depending on how `Pylint` is invoked the `Users/Daniel/...` is added as well. This normalises strings that refer to paths in the normal `pylint` output.

Reply:
Hmm yeah, I remember that we had something similar in the config and had to use abspath and relpath in the result : https://github.com/PyCQA/pylint/blob/main/tests/config/test_functional_config_loading.py#L10

Reply:
We seem to handle it here:
https://github.com/PyCQA/pylint/blob/a1df7685a4e6a05b519ea011f16a2f0d49d08032/pylint/testutils/configuration_test.py#L92

I don't have a particular preference, but this seems to work fine and I don't think we will need to expand these tests much later on. So this hacky way of substituting using regex would be fine imo.